### DISCUSSION TITLE
How to expand AFL++ to fuzz embedded device without any binary nor debug interface?

### DISCUSSION THREAD

**Discussion by skisssbb** [AFLplusplus/AFLplusplus#1170] [Q&A]
Hi,

i am currently working on my masterthesis. For my use case I have to fuzz an automotive hardware. The use case is that the software on the automotive hardware is unknown => no binary nor debug interface for instrumentation/unicorn mode.

I have a way to detect jump between basic blocks which will allow me to imitate the AFL++ instrumentation process. However, the amount of basic blocks is unknown. 

My question is whether there is an interface which I can implement to:
1. Send the request/input to the hardware
2. Wait for the response
3. Write the response somewhere (into a bitmap?) so AFL++ can calculate the coverage and inturn mutate/prioritize the next input

I found the "AFLplusplus/utils/afl_proxy/afl-proxy.c" but I'm not sure how to further implement this:
- How to report the coverage back to AFL Fuzzer by updating the basic block coverage in `__afl_area_ptr`. If a transition from Block 0x00 to Block 0x01 happens, should `__afl_area_ptr[1]++` be incremented by 1 to signalize the transition to AFL Fuzzer? (According to https://lcamtuf.coredump.cx/afl/technical_details.txt)
- How does afl-proxy work? Will the afl-proxy main method start again for each test case? is the test case written in the `buf` variable by the AFL agent each time?

I also found the interesting method called `update_bitmap_score` which utilizes the following 2 arrays:
- `afl->fsrv.trace_bits[]` - I haven't quite understood what this array is for. Is the relevant byte in this array incremented whenever a transition between specific basic blocks occured e.g Block 0x00 to Block 0x01 => `afl->fsrv.trace_bits[1]++`?
- `top_rated` is probably a list of queue entries aka. input which specific transition trigger
Should I implement/manipulate these 2 variables somehow for it to work for my use case?

Your help is very much appreciated. In case of successful implementation I would happily submit my code.

Beste regards

Quang Anh Dang
Technical University Darmstadt, Germany

**Comment by vanhauser-thc** [comment]
well yes, afl_proxy/afl-proxy.c is your best bet in such a situation.

>     1. Send the request/input to the hardware

you receive the input by stdin. so read from there and pass it on to the target with whatever the input channel is.

>     3. Write the response somewhere (into a bitmap?) so AFL++ can calculate the coverage and inturn mutate/prioritize the next input
> I have a way to detect jump between basic blocks which will allow me to imitate the AFL++ instrumentation process. However, the amount of basic blocks is unknown.

the amount of total blocks is not important, but what do you detect? just that a jump happened, or also the new address?
if you see the address just do something like this:

``` 
uint16_t prev_loc = 0;
uint32_t rip;

while(rip = next_basic_block()) {
uint16_t cur_loc = xxhash64(&rip, 4, 0x1234) & 0xffff;
__afl_area_ptr[cur_loc ^ prev_loc]++;
prev_loc = cur_loc >> 1;
}
```

if you have no addresses ... well then you fuzzing doesnt make much sense

>     * How does afl-proxy work? Will the afl-proxy main method start again for each test case? is the test case written in the `buf` variable by the AFL agent each time?

IMHO it is obvious:
while(__afl_next_testcase()) is the loop that happens for each new testcase.
and yes the data will be in buf, and len will be the length read.

> I also found the interesting method called `update_bitmap_score` which utilizes the following 2 arrays:

ignore that. concentrate on afl-proxy.c

**Comment by skisssbb** [comment]
First of all, thank you very much for the quick response.
```
the amount of total blocks is not important, but what do you detect? just that a jump happened, or also the new address?
if you see the address just do something like this:
```
I'm working on a concept to improve code coverage with side-channel analysis. According to state-of-the-art I should be able to detect jumps and the relative distance (in instructions) from the current block to the destination block. With that information, I can give each basic block somewhat of a virtual address. That should be enough to estimate the code coverage right?

**Comment by vanhauser-thc** [comment]
maybe. dunno. please make a concrete example :)

**Comment by skisssbb** [comment]
So I try to give each basic block a virtual ID. For example we have following transitions between Block A (0x0000) -> B(0x1234) -> C (0x5678). Then my detection program would give back to AFL++ an ordered list of block IDs [0x0000, 0x1234, 0x5678].
AFL++ can then do something like this right?

```
__afl_area_ptr[0x0000^0x1234]++;
__afl_area_ptr[(0x1234 >> 1) ^ 0x5678]++;
```

**Comment by vanhauser-thc** [comment]
well if you can give each block a unique virtual id then you are already there.

cur_loc := the virtual block id now entered

```
__afl_area_ptr[cur_loc ^ prev_loc]++;
prev_loc = cur_loc >> 1;
```
and before a new fuzz input run you init prev_loc with 0.

**Comment by skisssbb** [comment]
Thank you, I will try this out

**Comment by skisssbb** [comment]
Hi Vanhauser, I forgot to ask how the crash report works. How can I tell AFL in afl-proxy that a crash has occured?

**Comment by domenukk** [comment]
You could report an error this value as status from your loop, similar to this
https://github.com/AFLplusplus/unicornafl/blob/0d82727f2b477de82fa355edef9bc158bd25d374/afl-unicorn-cpu-inl.h#L356

(see https://github.com/AFLplusplus/unicornafl/blob/0d82727f2b477de82fa355edef9bc158bd25d374/afl-unicorn-cpu-inl.h#L161 for the value)

**Comment by vanhauser-thc** [comment]
the last int of a fuzz run that a targets writes to afl-fuzz is a return code.
`if (WIFSIGNALED(value)) { crashed();}`
you have to set those bits so that WIFSIGNALED matches true

**Comment by skisssbb** [comment]
Hi,

I have implemented this and tried to fuzz a simple fake python server of 5 block IDs.
![BlockIDGenServer](https://user-images.githubusercontent.com/20577889/129596739-9cd8c96d-7d3a-446a-89d6-80d31053fce9.png)
A crash will be signalized, when ever the value 66 (0x42) is sent in the first byte (required the statemachine being in state 0004).

However, while testing this, I got following ERROR from AFL:
```
[-] Hmm, your system is configured to send core dump notifications to an
    external utility. This will cause issues: there will be an extended delay
    between stumbling upon a crash and having this information relayed to the
    fuzzer via the standard waitpid() API.
    If you're just testing, set 'AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1'.
 
    To avoid having crashes misinterpreted as timeouts, please log in as root
    and temporarily modify /proc/sys/kernel/core_pattern, like so:
 
    echo core >/proc/sys/kernel/core_pattern
 
[-] PROGRAM ABORT : Pipe at the beginning of 'core_pattern'
         Location : check_crash_handling(), src/afl-fuzz-init.c:2188
```
So I set the option via
```
export AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1
```
AFL then found the crashes:
![image](https://user-images.githubusercontent.com/20577889/129598280-0ce4286d-ef52-4a8e-bb55-e80de0cd65a4.png)

However, when i examined the crash files via hexdump, I see that none of the given crash inputs would trigger a crash since the input doesn't start with (0x42)
```
bi@bi-virtual-realm:~/afl-fuzz/fuzz1/out/default/crashes$ hexdump id\:000000\,sig\:01\,src\:000001\,time\:887\,op\:havoc\,rep\:8 
0000000 4042 0064 38ff 7f40 0046               
0000009
bi@bi-virtual-realm:~/afl-fuzz/fuzz1/out/default/crashes$ hexdump id\:000001\,sig\:01\,src\:000001\,time\:887\,op\:havoc\,rep\:2 
0000000 0202 4200 3040 00ff                    
0000007
bi@bi-virtual-realm:~/afl-fuzz/fuzz1/out/default/crashes$ hexdump id\:000002\,sig\:01\,src\:000001\,time\:894\,op\:havoc\,rep\:16 
0000000 9d18 9d15 2015 9d18 2015 0025          
000000b
bi@bi-virtual-realm:~/afl-fuzz/fuzz1/out/default/crashes$ hexdump id\:000003\,sig\:01\,src\:000001\,time\:894\,op\:havoc\,rep\:2 
0000000 fff8 4041 251c 0046                    
0000007
bi@bi-virtual-realm:~/afl-fuzz/fuzz1/out/default/crashes$ hexdump id\:000004\,sig\:01\,src\:000001+000002\,time\:991\,op\:splice\,rep\:8 
0000000 141d 0100 0012 1707 0017               
0000009
bi@bi-virtual-realm:~/afl-fuzz/fuzz1/out/default/crashes$ hexdump id\:000005\,sig\:01\,src\:000004+000002\,time\:1072\,op\:splice\,rep\:8 
0000000 0019 1801 0900 001e                    
0000007
```

I went ahead and write the testcase whenever a crash is triggered in afl-proxy.c to a file and indeed got:

```
4280FF7F6430788D270600, 11
424030255B175BFFFFFF7F, 11
42407002004240302542475F, 12
422518FF00, 5
4242424242170A4242000100004225, 15
4200F540301840FF00001B00F5FF00001B00404200F5FF000A00404200F5FF0000, 33
4222482546, 5
424043343164AD, 7
4240300100, 5
4242424201, 5
426417171F170A, 7
424030253025FF, 7
4240FFE805, 5
42403025180200424030251147804025, 16
4200021916FFEF1917003E003E, 13
422A3111107F11107F2800175617EB000A107F280017FA0000FA5617EB000A107F2800175617EB008001, 42
424030250A25, 6
4240454017B7FF7F171717170AFFFF0E17172833D20000100000, 26
4250302518020056403035, 11
```

My question is how should I interpret the crashes found in the folder `./out/default/crashes` and can ignore the ERROR by just setting AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES Flag?

**Comment by vanhauser-thc** [comment]
If you do not want to disable core analysis (why would you not?) then you can set AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES.

the data in out/default/crashes is the one that crashes the target. but as you do your own signaling to tell that there is a crash you might just be doing that wrong (obviously as your proxy tool has the right crash data though alf-fuzz saves the wrong). so what then happens is that afl-fuzz thinks that the attempt before or after the crash is the data for the crash.

**Comment by skisssbb** [comment]
Hello again,

i have now implemented a Server which takes input from afl-proxy, then forwarded this input to SUT (Arduino) and read its *self-implemented* instrumentation information and forward this back to afl-proxy to be written to __afl_area_ptr.

My process:
AFL-Fuzzer => AFL-Proxy => Python Server => SUT (Arduino) over PCAN

However, AFL doesn't seem to run with following error:
```bash
bi@bi-virtual-realm:~/$ /home/bi/afl-fuzz/AFLplusplus/afl-fuzz 
-i /home/bi/PycharmProjects/basicblockIDGenerator/afl_fuzzer/input 
-o /home/bi/PycharmProjects/basicblockIDGenerator/afl_fuzzer/output 
-t 10000 -m 10000 -- /home/bi/afl-fuzz/AFLplusplus/utils/afl_proxy/afl-proxy
[+] Loaded environment variable AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES with value 1
afl-fuzz++3.14a based on afl by Michal Zalewski and a large online community
[+] afl++ is maintained by Marc "van Hauser" Heuse, Heiko "hexcoder" Eißfeldt, Andrea Fioraldi and Dominik Maier
[+] afl++ is open source, get it at https://github.com/AFLplusplus/AFLplusplus
[+] NOTE: This is v3.x which changes defaults and behaviours - see README.md
[+] No -M/-S set, autoconfiguring for "-S default"
[*] Getting to work...
[+] Using exponential power schedule (FAST)
[+] Enabled testcache with 50 MB
[*] Checking core_pattern...

[-] Hmm, your system is configured to send core dump notifications to an
    external utility. This will cause issues: there will be an extended delay
    between stumbling upon a crash and having this information relayed to the
    fuzzer via the standard waitpid() API.
    If you're just testing, set 'AFL_I_DONT_CARE_ABOUT_MISSING_CRASHES=1'.

    To avoid having crashes misinterpreted as timeouts, please log in as root
    and temporarily modify /proc/sys/kernel/core_pattern, like so:

    echo core >/proc/sys/kernel/core_pattern
[!] WARNING: Could not check CPU scaling governor
[+] You have 2 CPU cores and 2 runnable tasks (utilization: 100%).
[*] Setting up output directories...
[+] Output directory exists but deemed OK to reuse.
[*] Deleting old session data...
[+] Output dir cleanup successful.
[*] Checking CPU core loadout...
[+] Found a free CPU core, try binding to #0.
[*] Scanning '/home/bi/PycharmProjects/basicblockIDGenerator/afl_fuzzer/input'...
[+] Loaded a total of 1 seeds.
[*] Creating hard links for all input files...
[*] Validating target binary...
[*] Spinning up the fork server...
[+] All right - fork server is up.
[*] Target map size: 65536
[*] No auto-generated dictionary tokens to reuse.
[*] Attempting dry run with 'id:000000,time:0,orig:1_1'...
    len = 10, map size = 1, exec speed = 3911792 us
[+] All test cases processed.
[!] WARNING: The target binary is pretty slow! See docs/perf_tips.md.
[+] Here are some useful stats:

    Test case count : 1 favored, 0 variable, 0 ignored, 1 total
       Bitmap range : 1 to 1 bits (average: 1.00 bits)
        Exec timing : 3.91M to 3.91M us (average: 3.91M us)

[*] -t option specified. We'll use an exec timeout of 10000 ms.
[+] All set and ready to roll!

[-] Unable to communicate with fork server. Some possible reasons:

    - You've run out of memory. Use -m to increase the the memory limit
      to something higher than 10000.
    - The binary or one of the libraries it uses manages to create
      threads before the forkserver initializes.
    - The binary, at least in some circumstances, exits in a way that
      also kills the parent process - raise() could be the culprit.
    - If using persistent mode with QEMU, AFL_QEMU_PERSISTENT_ADDR is
      probably not valid (hint: add the base address in case of PIE)

If all else fails you can disable the fork server via AFL_NO_FORKSRV=1.

[-] PROGRAM ABORT : Unable to communicate with fork server
         Location : afl_fsrv_run_target(), src/afl-forkserver.c:1284
```

In the dry run step 
```
[*] Attempting dry run with 'id:000000,time:0,orig:1_1'...
    len = 10, map size = 1, exec speed = 3911792 us
```
I could see my server receiving, forwarding the message and responding with the block ID
```
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Sending: Timestamp:        0.000000        ID: 50b7    S                DLC:  3    27 25 f8
Response: 0000
Response: 
```

I have tried several things, however nothing seems to help..
- increasing timeout with -t
- increasing memory either to default -m 0 or to -m 1000 or to -m 10000

My concern is the WARNING:
```
[!] WARNING: The target binary is pretty slow! See docs/perf_tips.md.
```
since when I try to rewrite the server to send back a dummy default blockID of "0000" right away without forwarding this to the DUT to reduce the processing time => it seems to work.

Also setting `AFL_NO_FORKSRV=1` gets rid of the error, however, no testcase arrives at the Python Server which doesn't help my case either.

Does anybody know how to solve this?

**Comment by vanhauser-thc** [comment]
```
[*] Attempting dry run with 'id:000000,time:0,orig:1_1'...
    len = 10, map size = 1, exec speed = 3911792 us
[+] All test cases processed.
```
so general execution works.
I would add some debug + timing output to afl-proxy and python server, and use AFL_DEBUG=1 and -t 10000 to see the output during fuzzing and go from there.

**Comment by skisssbb** [comment]
Thank you for your quick response. Apparently my afl-proxy if my server returns an empty blockID for the bitmap. This happens when AFL sends **too short** test case or **too long**. Is there a way to tell AFL to send exact size input?

**Comment by vanhauser-thc** [comment]
no but you can do that in afl-proxy - just skip the testcase.

**Comment by bemeyvif** [comment]
@skisssbb is your work public? I'm looking for inspiration on how to fuzz an embedded ARM cortex-m4 system.
@vanhauser-thc do you think this is a good approach:
(exchange with that platform is only feasible with a serial line)
I was thinking to have the afl-proxy in the middle.
Since I have the source code of the embedded system,
can afl-clang-lto can be re-compiled for ARM cortex-m4 cross compilation (to have instrumentation)?
Or is the instrumentation source code not easily portable to ARM cortex-m4?
The afl-proxy would be responsible for injecting new input, detecting a crash or read out the coverage information and give afl-fuzz feedback.

Would having access to a JTAG or SWD connection between the afl-proxy and embedded target improve fuzzing in some way?
(I'm unfamiliar with the insides of JTAG/SWD but maybe these return something like an address register that could indicate that new edge has been reached?)

Thank you

**Reply by vanhauser-thc** [reply]
for your target I would recommend to just compile your target normally, and analyze jtag (or whatever) data in afl-proxy to fill the coverage map there.
instrumenting the embedded code itself is more likely to be difficult because how would that data be visible to the afl-proxy outside? if you can make this happen then that would be faster.

**Comment by domenukk** [comment]
If you have the source code there's no need for a proxy, just fuzz the target directly.

**Reply by bemeyvif** [reply]
@domenukk thank you. According to my understanding of AFL, one needs to bridge the afl-fuzz executable running on a PC and the firmware on the embedded target. I thought this is where a proxy comes in (how to recognize the target crashed, how to feed input, how to return coverage feedback).

If you have a link that describes your approach on fuzzing the target directly,  can you share it? I like to read more about it.

**Comment by domenukk** [comment]
You can either run afl-fuzz on the device (if it's *nix), or run the embedded target in qemu on the host machine (like in firmwire https://github.com/FirmWire/FirmWire) or in unicorn.
Else, you can also cross-compile parts of the target to x86 to fuzz them, a parser, for example.

**Reply by bemeyvif** [reply]
@domenukk thank you
- In this case it is not a unix embedded device (cortex-m4: limited speed, limited mem.). 
- Running in AFL qemu mode would only be a solution for the generic part of the code (e.g. the typical hardware sensor code would not be fuzzed).
Do you have experience running embedded firmware on qemu? Because I like to know if the AFL qemu extension for detecting new code paths/edge is also working for non-x86 non-AMD code? (I've been playing with the idea of trying out qemu but the hardware target is also not readily supported by current qemu).
- Cross-compile selective parts/modules of the whole stack might be an idea but is much work (changing makefiles, maybe source while I'm not the owner).

I was thinking of having a proxy setup like @skisssbb might have done, would make it easier to fuzz the embedded platform. Hence my question. But any other approach and suggestion, I like to look into and comment on.

**Reply by domenukk** [reply]
If you ask me, I'll always suggest to re-host the firmware, but I'm also biased since that's the main thing I know how to do.
The easiest(tm) solution is to fuzz parts of the firmware in [unicornafl](https://github.com/AFLplusplus/AFLplusplus/tree/stable/unicorn_mode): set up the state, then start execution from a place that you are interested in, and hook everything you don't need.
Full-system rehosting can be a lot harder. For this you may want to wait until libafl-qemu supports full-system or hack on https://github.com/AFLplusplus/LibAFL/pull/551 - it's totally doable, but it may be a lot more painful. See also firmwire I linked above, which does full-system rehosting for some cortex R blobs.
